src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1102,13 +1102,50 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     x.base.base.loc);
             }
         }
+        ASR::ttype_t *result_type = nullptr;
         ASR::ttype_t *operand_type = ASRUtils::expr_type(operand);
         ASR::ttype_t *logical_type = ASRUtils::TYPE(
-            ASR::make_Logical_t(al, x.base.base.loc, 4, nullptr, 0));
+            ASR::make_Logical_t(al, x.base.base.loc, 1, nullptr, 0));
         ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(al, x.base.base.loc,
                 4, nullptr, 0));
         ASR::expr_t *value = nullptr;
 
+        // Assign the return types separately for each case,
+        // so that we don't get a type mismatch error for assignment.
+        if (ASRUtils::is_integer(*operand_type)) {
+            if (op == ASR::unaryopType::Not) {
+                result_type = logical_type;
+            } else {
+                result_type = operand_type;
+            }
+        } else if (ASRUtils::is_real(*operand_type)) {
+            if (op == ASR::unaryopType::Not) {
+                result_type = logical_type;
+            } else if (op == ASR::unaryopType::Invert) {
+                throw SemanticError("Bad operand type for unary " +
+                    ASRUtils::unop_to_str(op) + ": " + ASRUtils::type_to_str_python(operand_type),
+                    x.base.base.loc);
+            } else {
+                result_type = operand_type;
+            }
+        } else if (ASRUtils::is_logical(*operand_type)) {
+            if (op == ASR::unaryopType::Not) {
+                result_type = logical_type;
+            } else {
+                result_type = int_type;
+            }
+        } else if (ASRUtils::is_complex(*operand_type)) {
+            if (op == ASR::unaryopType::Not) {
+                result_type = logical_type;
+            } else if (op == ASR::unaryopType::Invert) {
+                throw SemanticError("Bad operand type for unary " +
+                    ASRUtils::unop_to_str(op) + ": " + ASRUtils::type_to_str_python(operand_type),
+                    x.base.base.loc);
+            } else {
+                result_type = operand_type;
+            }
+        }
```

Comment:
Is it possible to merge `is_real` and `is_complex` in a single `if` branch? I see that their bodies are same.